qdr/_ranker.cc

```diff
@@ -0,0 +1,417 @@
+        ++nwords_vocab;
+        nwords += it->second.first;
+    }
+
+    n2p1 = double(nwords) + double(nwords_vocab) + 1.0;
+}
+
+QDR::~QDR() {}
+
+double QDR::get_idf(const std::string& word)
+{
+    /// Get log(total docs / doc count for word) == IDF
+    double doc_freq;
+    counts_t::iterator got = counts.find(word);
+    if (got != counts.end())
+    {
+        // this word is in the corpus
+        doc_freq = got->second.second;
+    }
+    else
+        doc_freq = 1.0;
+    return log((double) total_docs / doc_freq);
+}
+
+scores_t QDR::score(doc_t& document, doc_t& query)
+{
+    if (document.empty() || query.empty())
+        throw std::invalid_argument(
+            "Document and query both need to be non-empty");
+
+    // compute counts, doc length, etc now since they are used
+    // in multiple components
+    word_counts_t query_counts = count_words(query);
+    word_counts_t doc_counts = count_words(document);
+
+    double nwords_document = 0.0;
+    for (word_counts_t::const_iterator it_doc = doc_counts.begin();
+        it_doc != doc_counts.end(); ++it_doc)
+        nwords_document += it_doc->second;
+
+    // now the scores
+    return score_single(doc_counts, query_counts, nwords_document);
+}
+
+std::vector<scores_t> QDR::score_batch(
+    doc_t& document, std::vector<doc_t>& queries)
+{
+    if (document.empty())
+        throw std::invalid_argument(
+            "Document and query both need to be non-empty");
+
+    // compute document level counts once, then iterate through queries
+    word_counts_t doc_counts = count_words(document);
+
+    double nwords_document = 0.0;
+    for (word_counts_t::const_iterator it_doc = doc_counts.begin();
+        it_doc != doc_counts.end(); ++it_doc)
+        nwords_document += it_doc->second;
+
+    // iterate through queries, compute scores
+    std::vector<scores_t> ret;
+    ret.reserve(queries.size());
+    for (std::vector<doc_t>::iterator it = queries.begin();
+        it != queries.end(); ++it)
+    {
+        if (it->size() == 0)
+            throw std::invalid_argument(
+                "Document and query both need to be non-empty");
+        word_counts_t query_counts = count_words(*it);
+        ret.push_back(score_single(doc_counts, query_counts, nwords_document));
+    }
+
+    return ret;
+}
+
+scores_t QDR::score_single(const word_counts_t& doc_counts,
+        const word_counts_t& query_counts,
+        const double& nwords_document)
+{
+    scores_t scores;
+    scores.tfidf = tfidf(doc_counts, query_counts);
+    scores.bm25 = okapi_bm25(doc_counts, query_counts, nwords_document);
+
+    lm_scores_t lm_scores = lm(doc_counts, query_counts, nwords_document);
+    scores.lm_jm = lm_scores.jm;
+    scores.lm_dirichlet = lm_scores.dirichlet;
+    scores.lm_ad = lm_scores.ad;
+
+    return scores;
+}
+
+double QDR::tfidf(
+    const word_counts_t& doc_counts, const word_counts_t& query_counts)
+{
+    /// Compute TF-IDF similarity between the document and query
```

Comment:
Yea it's documentation.  I haven't been following any convention here but had mostly chosen following the signature since it looks kinda like a Python docstring.